ldsc/src/main/scala/MergeRegionsStage.scala

```diff
@@ -20,47 +23,141 @@ class MergeRegionsStage(implicit context: Context) extends Stage {
+    // Create map dataset -> partitions
+    parts.groupBy(_.dataset)
+      .map { case (dataset, s3Partitions) =>
+        dataset -> s3Partitions.map(_.partition)
+      }
+  }
+
   /** The _SUCCESS file maps to a key prefix with partitioned bed files.
     *
-    * Outputs will be the set of unique partition name prefixes. For example:
+    * Outputs will be the set of unique partition name prefixes:
     *
-    *   `<annotation>___<tissue>___<additional_partitions>`
+    *   `<annotation>___<tissue>___<biosample>___<dataset>`
     *
     */
   override val rules: PartialFunction[Input, Outputs] = {
     case partitionFiles(dataset) => {
-      val part = "/partition=([^/]+)/".r
-      val partitionNames = context.s3
-        .ls(s"${partitionFiles.prefix.commonPrefix}${dataset}/")
-        .flatMap(obj =>
-          part
-            .findFirstMatchIn(obj.key)
-            .map(_.group(1))
-        )
-        .distinct
+      val partitionNames: Seq[String] =  s3PartitionMap.get(dataset).toSeq.flatten
 
-      // if there are no partitions, then we can ignore it
-      partitionNames match {
-        case Nil => Outputs.Null
-        case _   => Outputs.Named(partitionNames: _*)
+      val matchedPartitions: Seq[String] = partitionsList.flatMap { partitionsToMatch =>
+        partitionNames.map(MergeOutput.toMergeOutput(_, partitionsToMatch).toOutput)
       }
+      // if there are no partitions, then we can ignore it
+      if (matchedPartitions.nonEmpty) {
+        Outputs.Named(matchedPartitions: _*)
+      } else Outputs.Null
     }
   }
 
   override def make(output: String): Job = {
-    new Job(Job.Script(resourceUri("mergeRegions.pl"), output, subRegion))
+    val mergeOutput = MergeOutput.fromOutputString(output)
+    new Job(Job.Script(
+      resourceUri("mergeRegions.pl"),
+      mergeOutput.subRegion,
+      mergeOutput.partitionMatch,
+      mergeOutput.partitionOutput
+    ))
   }
 
-  /** Before the jobs actually run, perform this operation.
+  /** No need to rm files because the same file path will be used for the output and so it will be replaced
     */
-  override def prepareJob(output: String): Unit = {
-    context.s3.rm(s"out/ldsc/regions/${subRegion}/merged/${output}/")
-  }
 
   /** Update the success flag of the merged regions.
     */
   override def success(output: String): Unit = {
-    context.s3.touch(s"out/ldsc/regions/${subRegion}/merged/${output}/_SUCCESS")
+    context.s3.touch(s"${MergeOutput.fromOutputString(output).outputDirectory}/_SUCCESS")
     ()
   }
 }
+
+case class S3Partition(dataset: String, partition: String)
+
+case class Partition(name: String){
+  override def toString: String = name
+}
+
+object Partition {
+  val Null: Partition = Partition("null")
+  val AnnotationPartition: Partition = Partition("annotation")
+  val TissuePartition: Partition = Partition("tissue")
+  val BiosamplePartition: Partition = Partition("biosample")
+  val DatasetPartition: Partition = Partition("dataset")
+}
```

Comment:
Apologies for how weird this is, but it works for now. There wasn't a particularly clean well to generalize the merging process. But after the Null state failed I realized there is a way to do this which requires a bit too much thought to tack on. It'll need to be a refactor.

To outline: the merge job will take a series of steps and at each step peel off the last bit and merge down. So initially, it would merge with all four. Then it will consume that output as input and peel off dataset to produce an annotation-tissue-biosample merge. Then it peels off the next to form annotation-tissue merge. Then annotation, and finally a single `all` merge. I don't think it is terrible hard to set up, but it'll require some thinking.